### PR TITLE
Prepare Python 3.11 support

### PR DESCRIPTION
**Pull request by csko**
## Why are these changes needed?

Preliminary PR for adding Python 3.11 support, mapping out various dependencies, fixing issues.

### Main changes

- Upgrade cython to 0.29.32
- Add CI/CD steps for 3.11 wheel
- Change cython code to not use `recursion_depth`
- Update cloudpickle to latest
- Use newer manylinux2014 which has python3.11
- Condition certain python packages in requirements.txt on <3.11 that don't yet have a 3.11 version

### Checklist:

- cython
  - [x] remove deprecated `recursion_depth`
  - [x] exc_type https://github.com/cython/cython/issues/4500
- [ ] package dependencies https://pyreadiness.org/3.11/
  - [ ] llvmlite https://github.com/numba/llvmlite/pull/869
  - [ ] numba https://github.com/numba/numba/issues/8304
  - [ ] pyarrow
  - [ ] scikit-learn
  - [ ] pydantic
  - [x] cloudpickle
  - [x] upgrade cython to 0.29.32
  - [ ] tensorflow https://github.com/tensorflow/tensorflow/issues/58032
  - [ ] torch https://github.com/pytorch/pytorch/issues/86566
  - [ ] miniconda https://github.com/conda/conda/issues/11170
    - [ ] `docker/base-deps/Dockerfile`
- [x] claim to support 3.11 in setup.py
- [ ] cicd
  - [ ] .buildkite/
    - [ ] .buildkite/pipeline.build.yml
  - [ ] ci/
    - [ ] ci/build/test-wheels.sh
    - [ ] ci/build/build-docker-images.py
  - [ ] release tests
  - [ ] docker/retag-lambda/python_versions.txt
  - [ ] download_wheels.sh
  - [ ] wheels
    - [ ] `python/build-wheel-macos.sh`
    - [ ] `python/build-wheel-windows.sh`
- [ ] Tests
  - [ ] pytest ray/serve/tests
  - [ ] python python/ray/serve/examples/echo_full.py
  - [ ] bazel test //:core_worker_test
  - [ ] bazel test --build_tests_only //:all
  - [ ] //python/ray/tests:test_pydantic_serialization https://github.com/tiangolo/fastapi/issues/5048
  - [ ] //python/ray/train:test_torch_utils
- [ ] Documentation
  - [x] installation.rst

Current status: 
Linux and mac wheels build in CICD. Docker images will come in a separate PR.

## Related issue number

#27881

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by csko** [comment]
Some success building locally:
```
$ pip install -e . --verbose
...
Successfully installed ray-3.0.0.dev0
(venv) csko@cskos-MacBook-Air python % python
Python 3.11.0 (main, Oct 26 2022, 19:06:18) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import ray
```

File: python/ray/cloudpickle/cloudpickle.py
Comment (csko) [reply python/ray/cloudpickle/cloudpickle.py:12]: yes, I used https://github.com/cloudpipe/cloudpickle/blob/master/cloudpickle/cloudpickle.py as the updated version and it has this line

Comment (csko) [reply python/ray/cloudpickle/cloudpickle.py:12]: I will change it back so it passes lint

File: python/ray/_raylet.pyx
Comment (rickyyx) [inline python/ray/_raylet.pyx:266]: Could you give me some pointers for this deprecation? Tried searching but couldn't find anything related.

Comment (csko) [reply python/ray/_raylet.pyx:266]: https://docs.python.org/3/whatsnew/3.11.html

> recursion_depth: removed, use (tstate->recursion_limit - tstate->recursion_remaining) instead.

Comment (rickyyx) [reply python/ray/_raylet.pyx:266]: Ah nvm - got it.

File: python/ray/cloudpickle/cloudpickle.py
Comment (rickyyx) [inline python/ray/cloudpickle/cloudpickle.py:69]: Is the cloudpickle upgrade necessary for 3.11 support? If not, maybe we could separate these changes in another PR.

File: ci/build/build-docker-images.py
Comment (rickyyx) [inline ci/build/build-docker-images.py:46]: Nice catch. I think this is orthogonal to the py3.11 change. Let's make a PR to fix this first?

File: python/ray/cloudpickle/cloudpickle.py
Comment (csko) [reply python/ray/cloudpickle/cloudpickle.py:69]: I believe `//python/ray/tests:test_pydantic_serialization` or some similar test fails with the old version. That test still fails but with a different error now.

File: python/requirements_test.txt
Comment (rickyyx) [inline python/requirements_test.txt:50]: should we do conditional on the python version as in `python/requirements.txt`? 
Could this be why the test `tests:test_pydantic_serialization` was failing?

File: python/ray/cloudpickle/cloudpickle.py
Comment (rickyyx) [reply python/ray/cloudpickle/cloudpickle.py:69]: Gotcha - commented on the pydantic dependency. 

Ideally, if this is not part of the python3.11 requirement, we should do this in another PR as well.

**Comment by rickyyx** [comment]
IIUC, we are not able to build python 3.11 wheels for mac as well, since our mac CI machines currently don't have python3.11 installed. The newest mac AMIs of 10.15 are amzn-ec2-macos-10.15.7-20221005-230420, which  I believed don't include python3.11 as well. 

cc @krfricke, @simon-mo  - Any thoughts on unblocking wheels built for 3.11 on mac?

File: python/ray/cloudpickle/cloudpickle.py
Comment (rickyyx) [reply python/ray/cloudpickle/cloudpickle.py:69]: Nvm - I think this needs to be part of the python3.11, I tested with the built wheel, and there's incompatibility with older cloud pickle source code.

File: .buildkite/pipeline.build.yml
Comment (rickyyx) [inline .buildkite/pipeline.build.yml:167]: Let's remove the docker build jobs for now, this is not a blocker to generate wheels.

File: ci/build/test-wheels.sh
Comment (rickyyx) [inline ci/build/test-wheels.sh:51]: I believe we never do test-wheels for linux - so the codepath never gets triggered for this. 
We could remove this change to keep it minimal.

File: ci/build/test-wheels.sh
Comment (rickyyx) [inline ci/build/test-wheels.sh:55]: Same as above.

File: ci/build/test-wheels.sh
Comment (rickyyx) [inline ci/build/test-wheels.sh:91]: Same as above.

File: ci/ci.sh
Comment (rickyyx) [inline ci/ci.sh:392]: Let's remove the `"${folder}.data/purelib/ray/__init__.py"` line below at line 418 which is causing issues actually.

File: docker/base-deps/Dockerfile
Comment (rickyyx) [inline docker/base-deps/Dockerfile:51]: I think we don't need to change this if we are not building dockers for now.

File: docker/retag-lambda/python_versions.txt
Comment (rickyyx) [inline docker/retag-lambda/python_versions.txt:6]: we could revert this docker change.

File: python/requirements_test.txt
Comment (rickyyx) [reply python/requirements_test.txt:50]: Let's still do the conditional as other dependencies for 3.11.

File: release/nightly_tests/stress_tests/state_api_app_config.yaml
Comment (rickyyx) [inline release/nightly_tests/stress_tests/state_api_app_config.yaml:9]: I am removing these from this PR, let's not touch this. 
https://github.com/ray-project/ray/pull/31339

File: python/build-wheel-windows.sh
Comment (rickyyx) [inline python/build-wheel-windows.sh:11]: we could skip windows wheel build for now.

File: python/build-wheel-macos.sh
Comment (rickyyx) [inline python/build-wheel-macos.sh:21]: let's skip mac wheel build for now.

**Review by rickyyx — COMMENTED** [review]
I think we are super close (great callout of the dependencies change, the recursion deprecation, and the cloud pickle change). I rebased your change on one of my test [PR](https://github.com/ray-project/ray/pull/31345) to test things out, and I think we could do the below in this PR: 

1. Bump the cython version 
2. Make sure at least Linux wheels are generated OK (we could work on the windows wheel and mac wheel later) 
3. Add the minimal test
4. Doc changes for the experimental wheel for linux. 

Let's postpone the below for future PRs: 
1. Windows/Mac wheels (those are not super urgent ,and there are some hard blockers, e.g. CI mac machines we use for wheel building are not shipped with python3.11 yet) 
  - Or we could also add the window wheel but it's not super critical. 
3. Docker images build (those could come later as part of the python 3.11 beta support). I ran into issues with the docker build which are probably due to deps. I don't think it's worth the effort to drill down now as most of the deps will gradually catch up, and ray docker images are not super critical in experimental support. 


Let's make sure:
1. The new minimal python test passes for py 3.11
2. The Wheels for linux is generated

What do you think?

File: python/ray/cloudpickle/cloudpickle.py
Comment (rickyyx) [reply python/ray/cloudpickle/cloudpickle.py]: EDIT: these not reformatting changes, I was wrong.

File: ci/build/build-docker-images.py
Comment (csko) [reply ci/build/build-docker-images.py:46]: https://github.com/ray-project/ray/pull/31385

File: ci/build/test-wheels.sh
Comment (csko) [inline ci/build/test-wheels.sh:99]: should I keep these?

File: ci/ci.sh
Comment (csko) [reply ci/ci.sh:392]: If I remove that line, isn't the function just checking if BUILDKITE_COMMIT/TRAVIS_COMMIT is set? Shouldn't we just remove the whole check?

File: ci/build/test-wheels.sh
Comment (rickyyx) [reply ci/build/test-wheels.sh:98]: I guess we should revert it as well? Since py311 wheel for mac will not be built.

Or feel free to leave it if you are able to get macos wheel built successfully. I wasn't able to do so for CI. (You will need to unblock the mac pipeline here):

image

Comment (csko) [reply ci/build/test-wheels.sh:98]: I saw that option, but I don't have permissions to start the Mac build. Looks like the build stalled after a while.

I have been working on a Mac and it does build locally.

File: doc/source/ray-overview/installation.rst
Comment (csko) [inline doc/source/ray-overview/installation.rst:76]: What's the process for adding this artifact? The LinkCheck check if failing

File: python/requirements_test.txt
Comment (csko) [reply python/requirements_test.txt:47]: I forgot what the error message was, but I believe we need a newer pydantic for 3.11. I haven't figured out why that test fails.

File: doc/source/ray-overview/installation.rst
Comment (rickyyx) [reply doc/source/ray-overview/installation.rst:76]: I believe we could ignore the soft failure - the wheel will get automatically uploaded to the URL if it's a CI job in master.

File: python/requirements_test.txt
Comment (rickyyx) [reply python/requirements_test.txt:47]: Looks like it's working now?

File: python/ray/_raylet.pxd
Comment (rickyyx) [inline python/ray/_raylet.pxd:72]: nit:I believe we could remove `int revursion_depth`.

File: .buildkite/pipeline.build.yml
Comment (rickyyx) [inline .buildkite/pipeline.build.yml:166]: Looks like the PR is ready to be marked as non-draft? Let's make sure the macOS pipeline also works as expected even though we are not adding macos wheel. 

I will get a committer for a review as well. 

nit: Let's delete those(commented codes) in this PR maybe.

File: ci/build/test-wheels.sh
Comment (rickyyx) [reply ci/build/test-wheels.sh]: Ah gotcha - I could do that. But yeah, I think it should be rather unlikely for mac build to regress.

File: .buildkite/pipeline.build.yml
Comment (rickyyx) [reply .buildkite/pipeline.build.yml]: Lovely - just started it.

**Comment by rickyyx** [comment]
So the wheel grep seems to work! Could you remove the other "commented out" dead code? And I think we should be able to merge it.

**Comment by csko** [comment]
- updated PR with a description of the major changes
- merged with upstream
- removed commented out code

**Comment by rickyyx** [comment]
Thanks so much. LGTM! 

A few notes on test for other reviewers: 
1. Lint check is soft failing since the wheel from master branch hasn't been uploaded, so the URL in the doc is not valid yet. This should be auto-resovled once the PR is merged and CI runs on the master. 
2. Previous mac run see here: https://buildkite.com/ray-project/oss-ci-build-pr/builds/8659#01858375-6bb5-4733-b638-4a1d2c5a25ea

**Comment by rickyyx** [comment]
Thanks so much. LGTM! 

A few notes on test for other reviewers: 
1. Lint check is soft failing since the wheel from master branch hasn't been uploaded, so the URL in the doc is not valid yet. This should be auto-resovled once the PR is merged and CI runs on the master. 
2. Previous mac run see here: https://buildkite.com/ray-project/oss-ci-build-pr/builds/8659#01858375-6bb5-4733-b638-4a1d2c5a25ea

**Review by rickyyx — APPROVED** [review]
~~Also - @csko could you update the PR discription (title) to reflect what we have left out in the PR and what have been included? Thanks!~~
Oh nvm, I think it's documented there.

**Review by krfricke — APPROVED** [review]
LGTM

File: doc/source/ray-overview/installation.rst
Comment (richardliaw) [inline doc/source/ray-overview/installation.rst:55]: can we avoid this "off by one" format? maybe just put 3.11 at the bottom (and label as experimental if necessary?)

**Review by richardliaw — CHANGES_REQUESTED** [review]

**Review by richardliaw — DISMISSED** [review]
Ricky, the table doesn't seem to render anymore

File: doc/source/ray-overview/installation.rst
Comment (csko) [inline doc/source/ray-overview/installation.rst:61]: I don't think this worked, let me fix

Comment (pcmoritz) [reply doc/source/ray-overview/installation.rst:61]: Also, Python 3.5 is not supported any more, let's remove that :)

Comment (csko) [reply doc/source/ray-overview/installation.rst]: fixed, and I also removed 3.5

**Review by pcmoritz — APPROVED** [review]
This looks good to me, but I did notice that the download link for the Python 3.11 wheels in the docs is currently not working. Will that be automatically fixed once the PR is merged or do we need to do anything else?

**Comment by pcmoritz** [comment]
I checked that wheels are only uploaded after the PR is merged -- let's try to merge it :)

**Review by richardliaw — APPROVED** [review]

**Review by wuisawesome — APPROVED** [review]

**Comment by pcmoritz** [comment]
After merging, the link is now indeed working 🎉

**Comment by grizzlybearg** [comment]
Are there any plans for windows python 3.11 wheels?

**Pull request merged**
